src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,284 @@
+            if (localType is null)
+            {
+                return default;
+            }
+
+            var initializer = (localDeclaration.Initializer ?? declarator.Initializer).Value;
+            if (initializer is null || initializer.Kind == OperationKind.Invalid)
+            {
+                return default;
+            }
+
+            var isValidSelection = await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                document,
+                selection,
+                node: declarationSyntax,
+                holes: ImmutableArray.Create(initializer.Syntax),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!isValidSelection)
+            {
+                return default;
+            }
+
+            if (!IsLegalUsingStatementType(semanticModel.Compilation, disposableType, localType))
+            {
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        private static bool IsLegalUsingStatementType(Compilation compilation, ITypeSymbol disposableType, ITypeSymbol type)
+        {
+            if (disposableType == null)
+            {
+                return false;
+            }
+
+            // CS1674: type used in a using statement must be implicitly convertible to 'System.IDisposable'
+            return compilation.ClassifyCommonConversion(type, disposableType).IsImplicit;
+        }
+
+        private async Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var (declarationStatement, localVariable) = await FindDisposableLocalDeclaration(document, span, cancellationToken).ConfigureAwait(false);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+
+            var statementsToSurround = GetStatementsToSurround(declarationStatement, localVariable, semanticModel, syntaxFactsService, cancellationToken);
+
+            // Separate the newline from the trivia that is going on the using declaration line.
+            var trailingTrivia = SplitTrailingTrivia(declarationStatement, syntaxFactsService);
+
+            var usingStatement =
+                CreateUsingStatement(
+                    declarationStatement,
+                    trailingTrivia.sameLine,
+                    statementsToSurround)
+                    .WithLeadingTrivia(declarationStatement.GetLeadingTrivia())
+                    .WithTrailingTrivia(trailingTrivia.endOfLine);
+
+            if (statementsToSurround.Any())
+            {
+                var parentStatements = GetStatements(declarationStatement.Parent);
```

Comment:
this is super subtle, and should be documented.  effectively, as i understand it, you won't have any statementsToSurround if hte parent is something like an if-statemnet.  that's why we don't end up throwing in the call to GetStatements.

Reply:
Does https://github.com/dotnet/roslyn/pull/30106/commits/ad919e8a4aea7a3cbe5c27ee6b2d193956737b85 help? Not sure what else to do that will be seen in all the places. Any ideas?

Reply:
See https://github.com/dotnet/roslyn/pull/30106#discussion_r229905424